core/src/main/java/com/datastax/oss/driver/internal/core/os/Native.java

```diff
@@ -25,8 +25,24 @@
 
   private static class LibcLoader {
 
+    /* These values come from Graal's imageinfo API which aims to offer the ability to detect
+    * when we're in the Graal build/run time via system props.  The maintainers of Graal have
+    * agreed that this API will not change over time.  We reference these props as literals
+    * to avoid introducing a dependency on Graal code for non-Graal users here. */
+    private static final String GRAAL_STATUS_PROP = "org.graalvm.nativeimage.imagecode";
+    private static final String GRAAL_BUILDTIME_STATUS = "buildtime";
+    private static final String GRAAL_RUNTIME_STATUS = "runtime";
+
     public Libc load() {
       try {
+        if (isGraalBuildTime()) {
+          LOG.info("Using Graal-native functions for Graal build time");
+          return new GraalLibc();
+        }
+        if (isGraalRunTime()) {
```

Comment:
I don't understand why do we need to create two conditions. 
If the graal is available in the build time, it is also available in the build time, isn't it?
If so we can coalesce those into one if that checks:  
```
return System.getProperty(GRAAL_STATUS_PROP).equals(    return System.getProperty(GRAAL_STATUS_PROP).equals(GRAAL_BUILDTIME_STATUS) || 
return System.getProperty(GRAAL_STATUS_PROP).equals(GRAAL_RUNTIME_STATUS)
```

Reply:
Yeah I tend to agree. Maybe if we needed different implementations at some point in the future, but that sounds far-fetched and we could always make the change later...

Reply:
My original thinking was that we'd need two checks to address the following cases:

* At build-time we want Graal's build-time analysis logic to follow that path rather than worry about the jnr path
* At run-time we wan to select the impl in order to access the underlying System/libc functionality

I don't think the run-time point above is terribly controversial... but as I reflect on it more I think the build-time requirement is bogus.  Built into my logic above is a belief that the build-time analysis done by the native image build is aware of these system properties and can detect branching logic based on them.  This is... somewhat silly.  A much more plausible explanation is that the analysis process just considers _both_ paths (realizing that the actual selection will be made at run-time).  As a result there's no real reason to worry about build-time at all.

Local testing seems to confirm this hypothesis, so I'll remove the build-time constraint all together.

Reply:
I think you need to leave the `build-time`. Without it, the quarkus app that is using JNR will not compile with incomplete classpath errors.

Reply:
It shouldn't result in incomplete classpath errors, but you do raise an interesting point @tomekl007 ... it's quite possible my sample app isn't doing build-time initialization of any of the native classes.  Which means we might not be exercising everything we need for testing here.

Yup, looks like that might be the case.  Quarkus does apparently trigger build-time init of these classes (via Uuids) which leads us to the following:

```
Error: Class initialization of com.datastax.oss.driver.api.core.uuid.Uuids failed. Use the option --initialize-at-run-time=com.datastax.oss.driver.api.core.uuid.Uuids to explicit
ly request delayed initialization of this class.                                                                                                    
Detailed message:                                                      
                                                                                                
com.oracle.svm.core.util.UserError$UserException: Class initialization of com.datastax.oss.driver.api.core.uuid.Uuids failed. Use the option --initialize-at-run-time=com.datastax
.oss.driver.api.core.uuid.Uuids to explicitly request delayed initialization of this class. 
Detailed message:                                                                                  
                                                                                              
        at com.oracle.svm.core.util.UserError.abort(UserError.java:75)                     
        at com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:221)                                                                                               at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:736)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:530)                                                                                       
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)                                                                                
        at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)                   
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)                                                                                                       
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)                                                                                           
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)               
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)                                                                                          
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Class initialization of com.datastax.oss.driver.api.core.uuid.Uuids failed. Use the option --initial
ize-at-run-time=com.datastax.oss.driver.api.core.uuid.Uuids to explicitly request delayed initialization of this class.                                                          
Detailed message:                                                                                                                                                                
                                                                                                                                                                                 
        at com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:126)                                                                        
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:733)
        ... 7 more                                                        
Caused by: java.lang.UnsatisfiedLinkError: com.datastax.oss.driver.internal.core.os.GraalGetpid.getpid()I
        at com.datastax.oss.driver.internal.core.os.GraalGetpid.getpid(Native Method)                                                                                            
        at com.datastax.oss.driver.internal.core.os.GraalLibc.getpid(GraalLibc.java:36)                                                                                          
        at com.datastax.oss.driver.internal.core.os.Native.getProcessId(Native.java:82)  
        at com.datastax.oss.driver.api.core.uuid.Uuids.getProcessPiece(Uuids.java:193)        
        at com.datastax.oss.driver.api.core.uuid.Uuids.makeNode(Uuids.java:163)             
        at com.datastax.oss.driver.api.core.uuid.Uuids.makeClockSeqAndNode(Uuids.java:226)                                              
        at com.datastax.oss.driver.api.core.uuid.Uuids.<clinit>(Uuids.java:116)      
        at sun.misc.Unsafe.ensureClassInitialized(Native Method)            
...
```

Looks like Uuids attempt to get at the PID is leading to a build-time attempt to access the native code which likely isn't available yet.

I've got an idea for how to work around this, need to do some more testing.